Migrate the Customers > Title pages (Only the list)

Progi1984 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Migrate the Customers > Title list (began by @tegbessou :heart_eyes:)
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #10628
| How to test?      | Make assets and Go to this URI: `/configure/shop/title`<br />You should see the list with Id, Social Type, Gender, Image, Edit, and Delete<br>You should be able to sort by Id, and Social Type<br>You should be able to filter by Id, Social Type, and Gender<br>
 Possible impacts? | No


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/28075)
<!-- Reviewable:end -->

Progi1984 committed:
Migrate the Customers > Title pages (Only the list)

Julievrz commented on src/Core/Grid/Definition/Factory/TitleGridDefinitionFactory.php:
```diff
@@ -0,0 +1,238 @@
+                    ->setTypeOptions([
+                        'required' => false,
+                        'attr' => [
+                            'placeholder' => $this->translator->trans('Search title', [], 'Admin.Shopparameters.Feature'),
```
```suggestion
                            'placeholder' => $this->translator->trans('Search title', [], 'Admin.Global'),
```

Just in case, it could be used elsewhere in the software.

Progi1984 replied:
Fixed :)

Julievrz replied:
Thanks!

jolelievre commented on src/Adapter/Title/TitleImageThumbnailProvider.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    public function getPath($titleId): ?string
+    {
+        $imageTag = $this->imageManager->getThumbnailForListing(
```
Strictly speaking, this does more than just return a path, internally the ImageManager creates an image file but all service implementing `ImageProviderInterface` and the naming "Provider" is confusing enough to allow generating the path AND the image file itself so it's fine

Maybe we should clarify this in the `ImageProviderInterface` though if you're up to it? Indicate that the method is used for both thumbnail AND path generation

Progi1984 replied:
Fixed.

jolelievre commented on classes/Gender.php:
```diff
@@ -31,6 +31,10 @@
 {
+    public const GENDER_MALE = '0';
+    public const GENDER_FEMALE = '1';
+    public const GENDER_NEUTRAL = '2';
```
I don't believe this last value is part of the fixtures data It means it should be the default use case when type is neither male nor female?

ttoine replied:
or, better: you add an other value "unknown", that becomes the default.

eshraw replied:
As discussed with the team :
- we keep the existing fixtures (Mr/Mme)
- when creating a new one --> “other” is set by default as the non-gendered option. "Other" replaces “neutral”
- “unknown” is unnecessary as it would essentially create a virtual double of “other” without any value since this gender field is tied to the fixtures.

jolelievre commented on src/Core/Grid/Data/Factory/TitleGridDataFactory.php:
```diff
@@ -0,0 +1,116 @@
+                case Gender::GENDER_FEMALE:
+                    $titles[$i]['type'] = $this->translator->trans('Female', [], 'Admin.Shopparameters.Feature');
+                    break;
+                case Gender::GENDER_NEUTRAL:
```
```suggestion
                default:
```

Should we consider this the default state? Any type not recognized as male or female is neutral?

Progi1984 replied:
Fixed. And validated with @eshraw

jolelievre commented on classes/Gender.php:
```diff
@@ -31,6 +31,10 @@
  */
 class GenderCore extends ObjectModel
 {
+    public const GENDER_MALE = '0';
+    public const GENDER_FEMALE = '1';
+    public const GENDER_NEUTRAL = '2';
```
Suggestion:

```suggestion
    public const GENDER_TYPE_MALE = '0';
    public const GENDER_TYPE_FEMALE = '1';
    public const GENDER_TYPE_NEUTRAL = '2';
```

To emphasize the fact that it's linked to the type field

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Grid/Query/TitleQueryBuilder.php:
```diff
@@ -0,0 +1,138 @@
+                continue;
+            }
+
+            if ($filterName === 'id_gender') {
```
I think `type` should be tested as strict equality as well as it's a tinyint configuration

Progi1984 replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+     *     redirectRoute="admin_title_index",
+     *     message="You need permission to delete this."
+     * )
+     * @DemoRestricted(redirectRoute="admin_title_index")
```
Never knew what's the use of these Demo attributes 😅

Progi1984 replied:
Check here `src/PrestaShopBundle/Security/Annotation/DemoRestricted.php`.

jolelievre reviewed (commented):
Thanks @Progi1984 and @tegbessou 

Overall everything is good from a grid perspective, I just have a few questions and suggestions regarding the Gender domain itself

NeOMakinG reviewed (dismissed):
front lgtm

Progi1984 commented:
@Julievrz Validated with @eshraw, there is not anymore "Neutral" gender type but "Other" gender type.

jolelievre reviewed (dismissed):
Thanks @Progi1984

Amoifr commented on src/Adapter/Title/TitleImageThumbnailProvider.php:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * @var string
+     */
+    private $genderDir;
```
genderDir is an enough explicit name ? genderImageDirectoryName ?

Progi1984 replied:
Fixed

Amoifr commented on src/Core/Grid/Query/TitleQueryBuilder.php:
```diff
@@ -0,0 +1,138 @@
+    private function getTitleQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $builder = $this->connection->createQueryBuilder()
+            ->from($this->dbPrefix . 'gender', 't')
```
copy/paste ? gender is aliased as "t" ... is syntaxly correct but not very relevant

Progi1984 replied:
Fixed

Amoifr commented on src/Core/Grid/Query/TitleQueryBuilder.php:
```diff
@@ -0,0 +1,138 @@
+    {
+        $builder = $this->connection->createQueryBuilder()
+            ->from($this->dbPrefix . 'gender', 't')
+            ->innerJoin('t', $this->dbPrefix . 'gender_lang', 'tl', 't.id_gender = tl.id_gender')
```
also gender_lang as tl ...

Progi1984 replied:
Fixed

NeOMakinG reviewed (dismissed):
FE lgtm

Amoifr commented on src/Core/Search/Filters/TitleFilters.php:
```diff
@@ -0,0 +1,55 @@
+     */
+    public static function getDefaults()
+    {
+        return [
```
any betterment to get it from a service that retrieve it from the conf ?

Progi1984 replied:
I think it's a improvement we could do in an another PR for applying that to all filters.

Amoifr commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+     */
+    public function indexAction(Request $request, TitleFilters $filters): Response
+    {
+        $titleGridFactory = $this->get('prestashop.core.grid.factory.title');
```
dependency injection .... what about to get it from an injection by method parameters ?

Progi1984 replied:
I think it's a improvement we could do in an another PR for applying that to all grids for consistency.

Amoifr commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+                true,
+                [],
+                [
+                    'addgender' => '',
```
camelcase : no
snakecase : no
;-)

Progi1984 replied:
Normal, it's the legacy action name.

Amoifr commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+                true,
+                [],
+                [
+                    'updategender' => '',
```
camelcase, snakecase , other ?

Progi1984 replied:
Normal, it's the legacy action name.

Amoifr commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+     *
+     * @return Response
+     */
+    public function editAction(int $titleId): Response
```
titleId or genderId. One thing = one name

Progi1984 replied:
Gender is the old table name.
But the good name is Title.

Amoifr commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleController.php:
```diff
@@ -0,0 +1,155 @@
+     */
+    public function deleteAction(int $titleId): RedirectResponse
+    {
+        return $this->redirect(
```
$this->redirectToRoute('routename'); is not applicable ?

Progi1984 replied:
Not because I haven't migrated at the moment the delete action (and all actions).

Amoifr commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/title.yml:
```diff
@@ -0,0 +1,53 @@
+  defaults:
+    _controller: 'PrestaShopBundle\Controller\Admin\Configure\ShopParameters\TitleController::indexAction'
+    _legacy_controller: AdminGenders
+    #_legacy_link: AdminGenders
```
I think it missed a space after the # to respect the syntax.... but not sure

florine2623 reviewed (commented):
Hello @Progi1984 @tegbessou ,

Thanks for the PR !

- What is the name of the second column ?
In your PR ticket, it is `Social Type`
In the Shop in /configure/shop/title, it is `Title`
And it the old page, it is `Social Title`
What is the correct one ? I couldn't find any specs or mockup on this.

- I can't bulk delete my social titles 🚫 (Deletion one by one is fine.)
 
https://user-images.githubusercontent.com/16019289/162962905-6ed8796a-e6ae-4a94-baa0-204452d66670.mov

- Pagination works well
- Filtering by ID or Title doesn't work. I have to choose a Gender 🚫 

https://user-images.githubusercontent.com/16019289/162963819-145ae7eb-2b80-4ba6-8a31-7ea1e44d4faf.mov

- Filtering by Gender and Resetting is OK
- Sorting by ID and Titles is OK
- Also, the 2 links redirect to the old page, is it the wanted behavior ?
<img width="387" alt="Screenshot 2022-04-12 at 14 39 08" src="https://user-images.githubusercontent.com/16019289/162964319-85aed269-df43-43e2-8d50-650047223714.png">

Could you check please ?
Thanks!

Progi1984 commented:
@PrestaShop/prestashop-core-developers & @florine2623 : 
- Fixed the title of the second column.
- The bulk delete will be migrated in a future PR.
- Fixed the filtering (the type is not anymore required)
- Two links will be migrated in the future, not in this PR.

florine2623 reviewed (commented):
Thanks for the details @Progi1984 ,

Looking good to me then.

I just have one more question regarding filtering on ID, I need confirmation from @PrestaShop/product-team for this.
On old page I have :
<img width="1208" alt="Screenshot 2022-04-14 at 14 44 59" src="https://user-images.githubusercontent.com/16019289/163393843-d7f78767-dcef-40a4-a43f-0eab57ddf013.png">
On migrated page, I have :
<img width="1213" alt="Screenshot 2022-04-14 at 14 45 34" src="https://user-images.githubusercontent.com/16019289/163393855-29cc504b-00fc-45b6-9fcd-d5c8ec2c0c71.png">

What should be the right behavior ? 

Thanks!

Progi1984 commented:
@florine2623 Just fixed this bug.

Progi1984 commented:
@NeOMakinG  & @PierreRambaud : Feedbacks from @florine2623 adressed. Please review.

matks commented on tests/Integration/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/TitleControllerTest.php:
```diff
@@ -0,0 +1,63 @@
+use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
+use Symfony\Component\HttpFoundation\Response;
+
+class TitleControllerTest extends WebTestCase
```
I'm not a fan of these tests (nothing new here) but as it's small and I have nothing better ✅  for me

Progi1984 replied:
Thanks

jolelievre reviewed (dismissed):
Thanks @Progi1984

florine2623 reviewed (commented):
Hello @Progi1984,

I have an exception when I try to filter by ID.

https://user-images.githubusercontent.com/16019289/165470892-d90377f4-c5f8-446b-b2e3-5d12adb7caa7.mov

<img width="1068" alt="Screenshot 2022-04-27 at 09 58 09" src="https://user-images.githubusercontent.com/16019289/165470874-29dc11bb-e669-49ce-80ea-2d076e248335.png">

Filter by Social Title and Gender is OK.
Could you check please ?

Progi1984 committed:
Fixed filters with disambiguation in the query

NeOMakinG reviewed (approved):
front LGTM

jolelievre reviewed (approved):
Thanks @Progi1984

florine2623 reviewed (approved):
@Progi1984 ,

Filter, sort and display of Titles work well !
It is QA ✅

Progi1984 commented:
Thanks @florine2623
